### ISSUE TITLE
v4 getSignedUrl with Generation failing with SignatureDoesNotMatch

### ISSUE THREAD

**Issue by stuvet** [googleapis/nodejs-storage#953]
##  Environment details
  - OS: Linux 5.3.10-arch1-1
  - Node.js version: 12.3.0
  - npm version: 6.12.1
  - `@google-cloud/storage` version: 4.1.3

#### Steps to reproduce
  0. Prepare private bucket with versioning.
`
gsutil mb -c nearline gs://my-bucket
`
`
gsutil versioning set on gs://my-bucket
`
  1. Save file to storage bucket.
`
gsutil cp file.pdf gs://my-bucket/path/to/file.pdf
`
 2. Confirm upload & get object generation
`
gsutil ls -a gs://my-bucket/path/to/file.pdf 
`
`
 // gs://my-bucket/path/to/file.pdf#generation
`
  2. Create a **v2** signedUrl specifying a generation and download file via curl/browser => **success**
  3. Create a **v4** signedUrl specifying a generation and download file via curl/browser => **failure**

```
// fails on v4 whether generation is string or numeric. 
// Succeeds on v2. Both v4 & v2 succeed when generation is undefined.

const version = 'v4';
const bucketName = 'my-bucket';
const filename = 'path/to/file.pdf';
const generation = 1575497958681459;

const [url] = await storage
              .bucket(bucketName)
              .file(filename, {generation: generation})
              .getSignedUrl({
                  version: version,
                  action: 'read' ,
                  expires: Date.now() + 2 * 60 * 1000,
                });
```
#### Error Message
```
code=SignatureDoesNotMatch

message=The request signature we calculated does not match the signature you provided. Check your Google secret key and signing method.

StringToSign=GOOG4-RSA-SHA256 20191205T153311Z 20191205/auto/storage/goog4_request 4397be36b625f65c72de14ef8656ba3686c54a8e21af3778de61afd60eb56ca7

CanonicalRequest=GET /my-bucket/path/to/file.pdf X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=firebase-adminsdk-29ulz%40my-project-XXXXXX.iam.gserviceaccount.com%2F20191205%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20191205T153311Z&X-Goog-Expires=121&X-Goog-SignedHeaders=host&generation=1575497958681459 host:storage.googleapis.com host UNSIGNED-PAYLOAD
```
Both v2 and v4 work fine when generation is undefined.

I have looked through the other pending issues and can't find mention in #912. It's possible this is related to URL encoding of slashes in the object name, though this was fixed in #905

I'm sticking to v2 for now as it works as expected.

Thanks!

**Comment by dmitry-fa** [comment]
v4 getSignedUrl problem with Generation exists not only in nodejs storage, but in java as well: [#7053](https://github.com/googleapis/google-cloud-java/issues/7053)
Perhaps, it should be resolved on the server side first.

**Comment by dmitry-fa** [comment]
backend issue: https://issuetracker.google.com/issues/146802530

**Comment by jkwlui** [comment]
This is indeed a bug in the client in that query parameters like `generation` wasn't included in part of the signature. We'll be addressing this as part of #1046, in which we will make sure all query parameters (user-provided, or set via an option like `generation`) are indeed signed.

Thanks for reporting the issue @dmitry-fa, and I appreciate your patience with this.

**Comment by frankyn** [comment]
I think this is now fixed, @jkwlui, can you confirm?

**Comment by jkwlui** [comment]
Fixed via @google-cloud/storage@v4.5.0
